### DISCUSSION TITLE
How to publish with Visual Studio 2019/2022

### DISCUSSION THREAD

**Discussion by rrdgcoks** [OrchardCMS/OrchardCore#11462] [Q&A]
Hello! Thanks in advance.

I have a working Orchard Core CMS site built locally using Visual Studio 2022. It's connected to a SQL Server database. I'm ready to test publishing. If anyone has pointers or links to documentation, I'd appreciate it. I couldn't find anything in the current Orchard Core docs.

Using Visual Studio, I tried to publish like you would a .NET web app. I went to "Publish" although not listed as an option in the Solution Explorer. I chose to publish to a folder. I added the folder and so forth to create a publishing profile. It was "successful" but the web server folder now has 2 files: a NUPKG file and a SNUPKG file. 

Please advise on how to publish to a web server. 

Thank you!

**Comment by Skrypt** [comment]
It works just like any other website publishing. You can use `Visual Studio Code` publish task to publish as I do. We don't provide any publishing profile because it is per need.

**Comment by rrdgcoks** [comment]
Thanks - I tried publishing with VS 2022 as normal but it produced NuGet packages in that folder and no launchable codebase. I'll try again.

**Reply by Skrypt** [reply]
That's because you did a `dotnet nuget publish` instead of a `dotnet publish`.
Refer yourself to Microsoft documentation here. This is general knowledge.

**Comment by rrdgcoks** [comment]
Actually I used Build -> Publish Selection, not command line. I can try using command line.
I noticed there was no connection string or database context previously, so I'm looking into adding string as a Service Dependency.

**Comment by rrdgcoks** [comment]
Which exact folder should I be trying to publish? What I'm seeing in Solution Explorer:

Solution 'OrchardCore'
- /build
- /docs
- /Solution items
- /src
- /Templates
- /test

**Reply by Skrypt** [reply]
build

**Comment by rrdgcoks** [comment]
`dotnet publish OrchardCore.sln --configuration Release --runtime win10-x86 --output \\myserver\myfolder --self-contained true`

This seems to have done the trick. The code looks right now in the server folder. I have a 403, but that's a separate issue. Thanks for the help.

**Reply by jtkech** [reply]
Normally you only need to do a publish on your web app, not on the entire solution, idem under VStudio.

**Comment by rrdgcoks** [comment]
Could you please tell me where my web app would be found in the file structure? I don't think /build looks right but I could be wrong.

Solution 'OrchardCore'

/build
/docs
/Solution items
/src
/Templates
/test

**Reply by jtkech** [reply]
E.g. in our repo /src/OrchardCore.Cms.Web

Then if you use the dotnet cli, first go under the above folder to execute the command

**Comment by rrdgcoks** [comment]
Based on what you said, I navigated to the /src folder in CLI and tried this:
`dotnet publish OrchardCore.Cms.Web --configuration Release --runtime win10-x86 --output \\server\wwwroottest\folder --self-contained true`

I got an error: `error NETSDK1129: The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, you must specify the framework for the published application.`

Based on this error, it would seem I need to include the -f or --framework flag. Would that be something like this? `--framework OrchardCore`

**Reply by jtkech** [reply]
If the current directory is OC.Cms.Web you don't need to specify it
We now only target net6.0, otherwise yes -f net6.0 as I remember
Then it depends on your server to use or not e.g. self contained.

**Reply by rrdgcoks** [reply]
I figured out how to Publish the OrchardCore.Cms.Web folder using Right Click->Publish. From there, I just needed to create a new profile. I think this is all I needed in order to publish. I did experience a Build Failed error, but I think I'm getting somewhere now. I'm still getting an error: `The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, you must specify the framework...". I'll just Google this error.

I'm not sure why my local installation is connected to a database but there was not database connection shown in the publishing. I'm going to have to research how Orchard handles the db connection / connection string. Or maybe I can just do the setup in the browser, like I did locally, except on the hosted/published deployment?

**Reply by rrdgcoks** [reply]
I am making progress, so I really appreciate your help.

The build failed, yet I loaded the website semi-successfully. It's not pulling some images correctly, but it appears to be working? I'm not sure what the issue is with target framework. I can try different publishing options.
